feat: improve keybind management

JulienZD opened the pull request:
Part of #30

* Added a `useKeybinds` hook which provides a `register` callback where you an register keybinds for that component (with the option to unregister on unmount, for modals)
* Keybinds for the active view are displayed globally (inspired by Lazygit)

JulienZD committed:
feat: manage keybinds globally

The BoardView has been converted as an example

JulienZD committed:
feat: support keybind aliases

JulienZD committed:
feat(keybinds): view specific keybind display

JulienZD committed:
feat: global keybinds display

JulienZD committed:
feat: handle known aliases

JulienZD committed:
feat: convert user select modal to use new keybinds

JulienZD committed:
fix(keybinds): prevent duplicate keybinds for reopened modals

RobinHeidenis commented on src/atoms/active-view.atom.ts:
```diff
@@ -0,0 +1,15 @@
+  },
+);
+
+const prevViewStoreAtom = atom<string | null>(null);
```
can you put this one above the previousViewAtom? Now we're kind of using it before it's declaration

RobinHeidenis commented on src/hooks/use-keybinds.ts:
```diff
@@ -0,0 +1,71 @@
+  }, [registerFn, actualCallback, ...deps]);
+};
+
+function hash(keybind: Keybind) {
+  return JSON.stringify({
+    key: keybind.key,
+    modifiers: keybind.modifiers ?? [],
+  });
+}
```
will this go well? I think we might need to sort the modifiers, imagine for example i have a `ctrl + shift + d` shortcut to delete something or whatever. Imagine in another component i accidentally write it as `shift + ctrl + d`, then it won't match will it?

RobinHeidenis commented on src/modals/select-users-modal.tsx:
```diff
@@ -1,57 +1,127 @@
+    { view: "SelectUsersModal", unregister: true },
+    (register) => {
+      register({
+        key: "j",
+        hidden: true,
+        name: "Down",
+        aliases: [
+          { key: "", modifiers: ["downArrow"] },
+          { key: "n", modifiers: ["ctrl"] },
+        ],
```
maybe we can make a definition somewhere for these common keybinds, that we can just do `...up` or something

JulienZD replied:
Good one, had it in mind when I saw the other keybinds

RobinHeidenis commented:
is there a way to make a view "exclusive"? For example, if i have the issue modal open, and i change the priority for example (so the priority modal is now open above the issue modal), I don't want to be able to open the change assignee modal. Is there any way to make a view exclusive so only that view's keybinds are active until closed?

RobinHeidenis commented on src/hooks/use-keybinds.ts:
```diff
@@ -0,0 +1,71 @@
+  deps: unknown[],
+) => void;
+
+export const useKeybinds: UseKeybinds = (options, callback, deps) => {
```
is there a way to make a view "exclusive"? For example, if i have the issue modal open, and i change the priority for example (so the priority modal is now open above the issue modal), I don't want to be able to open the change assignee modal. Is there any way to make a view exclusive so only that view's keybinds are active until closed?

JulienZD replied:
This should already be the behavior due to `options.view` -  the `activeKeybindsAtom` only grabs those of the active view, which is used by the global key handler

JulienZD committed:
refactor: minor feedback

JulienZD committed:
refactor: extract common keys + aliases

JulienZD committed:
fix(keybinds): ignore keybinds while search is active

JulienZD committed:
feat: convert select modal to use new keybinds system

JulienZD committed:
feat: move Board keybinds to BoardView

Left the navigation for now, might require a separate solution

JulienZD committed:
feat: move ViewIssueModal keybinds to new systen

JulienZD committed:
feat: re-add support for changing issue fields when viewing an issue

RobinHeidenis commented on src/board-view.tsx:
```diff
@@ -120,15 +86,107 @@ export const BoardView = () => {
+        ...CONFIRM_KEY,
+        name: "View issue",
+        hidden: true,
+        when: () => store.get(boardSearchStateAtom) !== "active",
```
maybe we can add a check here that there is actually an issue selected (amount of issues > 0). It goes well now, i think there's a check somewhere in the code but might be nice to make sure here

RobinHeidenis replied:
probably should do that for all the issue related modals

RobinHeidenis commented on src/board.tsx:
```diff
@@ -191,49 +175,24 @@ export const Board = ({
-      openIssueInBrowser(issue?.key ?? "");
-    } else if (input === "j" || key.downArrow) {
+    // TODO: not ideal that these are in their own useInput, but might be necessary due to the moving of the board, etc.
+    // Maybe extract to a separate `useKeyboardNavigation` hook with callbacks for each direction?
```
do we do this in many other places? If it's just here then i'd say it's probably fine for now like this, although we should see if we can improve the math down here, it looks kinda complicated

JulienZD replied:
Only here

RobinHeidenis commented on src/modals/select-lane-modal.tsx:
```diff
@@ -11,16 +11,18 @@ import { SelectModal } from "./select-modal.js";
 
 export const SelectLaneModal = ({
   onClose,
+  issueId: _issueId,
```
maybe rename `_issueId` to `issueIdOverride` or something? maybe `requestedIssueId`, depending on the use case

RobinHeidenis commented on src/modals/update-assignee.modal.tsx:
```diff
@@ -6,15 +6,19 @@ import { highlightedIssueAtom } from "../atoms/highlighted-issue.atom.js";
-export const UpdateAssigneeModal = () => {
-  const highlightedIssue = useAtomValue(highlightedIssueAtom);
+export const UpdateAssigneeModal = ({
+  issueId: _issueId,
```
here too

RobinHeidenis commented on src/modals/view-issue-modal.tsx:
```diff
@@ -1,35 +1,35 @@
+const topOffsetAtom = atom(0);
 
-  const description = new ADFRenderer(119, 33).render(
+const issueDescriptionAtom = atom((get) => {
+  const issue = get(issueAtom);
+  if (!issue) {
+    return [];
+  }
+
+  return new ADFRenderer(MAX_LINE_WIDTH, MINIMUM_LINES).render(
```
this appears to be causing some issues, where the description of a previous issue is still shown for a bit before showing the current one. See video

https://github.com/user-attachments/assets/39b6e403-274d-48fc-95c1-5c5c18496f24

RobinHeidenis replied:
i think the video is screwed up, but you can see it happen clearly at 4 seconds

JulienZD replied:
Noticed that too, how would you recommend fixing that? I think it's because of the issue atom being set in the useEffect, causing the issue atom to be out of sync with the query result every so slightly. 

This is also the case in other places I think, where we have to set the atom in a useEffect 

I wonder whether just using the issue atom for the ui as well, rather than the hook result will fix it. Probably have to empty the issue atom on unmount too though, so the `!issue` case gets hit

JulienZD replied:
Nvm, it's as easy as just resetting the issue atom on unmount

RobinHeidenis replied:
This also seems to be a problem with the select lane modal

JulienZD committed:
refactor: improve naming

JulienZD committed:
fix: only act if there's a highlighted or opened issue

JulienZD committed:
fix: reset issue when closing view modal

This prevents a brief flash of the previously opened issue description when opening
another issue

RobinHeidenis committed:
refactor: remove issueDescriptionAtom
